### DISCUSSION TITLE
Unfolding Repeats

### DISCUSSION THREAD

**Discussion by jonaro00** [CoderLine/alphaTab#1049] [Ideas]
Making this thread for discussing this feature. Idea from in https://github.com/CoderLine/alphaTab/discussions/671#discussioncomment-1587693. In short: A setting that "unfolds" all repeats and alternate endings, thereby making the sheet look exactly like it is played.

To start off, I propose this setting can be called `settings.display.unfoldRepeats`, and be a boolean, default false.

There are many concerns and interactions to discuss:

- [ ] Preprocessing of original score to an unfolded score:
  - [ ] In which layer should this logic exist?
  - [ ] Should it be done in advance or only when turning on this setting? (I'm guessing the latter). 
- [x] ~~`display.startBar` and `display.barCount`: Should unfolding allow showing only (parts of) specific laps in a repeat (for example only one variant of the alternate endings)? In that case a different start offset calculation might be needed.~~ Let the developer handle it.
- [ ] _There is probably more to consider..._

**Comment by Danielku15** [comment]
Would be a valuable addition that is supported already by quite some music notation software (e..g MuseScore and Soundslice). 

>  In which layer should this logic exist?
>  Should it be done in advance or only when turning on this setting? (I'm guessing the latter).

I would say the easiest way to implement the feature would be to rewrite the data model after loading. This way for the rest of alphaTab it looks almost like it is simply a score without repeats. The `MidiPlaybackController` would help to build the new score with the right bar order. The "con" about this solution is that folding/unfolding is not a dynamic operation. Usage of this feature would be a bit more complex because you would load the score, call the component to "clone/rewrite" it to an unfolded score and put this into alphaTab. To "fold" again, the original score would be put into alphaTab again. 

Having it on a different layer would likely need quite some rework in almost all components. At many places there is the assumption of having one BarRenderer for one Bar and we assume a strong relation about next/previous bars in the model also be like this in the display. (e.g. when finding the render location for inter-connected notes through slurs). Beside that a lot of components providing visual locations by a given bar/note would need to be transformed into arrays. And that's just the start. 

> display.startBar and display.barCount:...

With the approach above, it would be up to the developer using the feature to correctly specify the bars according to the folder or unfolded version. 

>  There is probably more to consider...

Cosidering a stepwise approach of such a feature it might be easiest to provide a new component in alphaTab for doing the unfolding and then provide some concrete example how it would be used to provide a folding/unfolding feature for end-users. This will show us already what complexities might be involved and potentially can or cannot be solved in alphaTab itself. 

I think at some point we could move some more features into alphaTab so that it holds a folded and unfolded version of the score (when the feature is activated) and do certain back&forth mappings on the public API surface.

**Comment by jonaro00** [comment]
> rewrite the data model after loading

Do you mean modifying the model in place or generating an unfolded version to be stored in parallel? Should a regeneration be done initially, after first render, or the first time the unfold setting is changed?

> Cosidering a stepwise approach of such a feature it might be easiest to provide a new component in alphaTab for doing the unfolding

Or perhaps it could be developed on a different branch.

Is a rough starting point to have a second `score` variable in AlphaTabApiBase and switch between them?
